File: LC/[0841] Keys and Rooms - iterative.py
# class Solution:
#     def canVisitAllRooms(self, rooms: list[list[int]]) -> bool:
#         stack=[0]
#         visited={0}

#         while stack:
#             room = stack.pop()
#             visited.add(room)
#             for kk in rooms[room]:
#                 if kk not in visited: stack.append(kk)
        

#         return len(visited) == len(rooms)
            

class Solution:
    def canVisitAllRooms(self, rooms: list[list[int]]) -> bool:
        stack = [0]
        visited = {0}

        while stack:
            room = stack.pop()
            for kk in rooms[room]:
                if kk not in visited:
                    visited.add(kk)
                    stack.append(kk)

        return len(visited) == len(rooms)

"""
# we want to practice implement the DFS by iterative implementation
"""
        

File: LC/[0841] Keys and Rooms - iterative.java
class Solution {
    public boolean canVisitAllRooms(List<List<Integer>> rooms) {
        Deque<Integer> stack = new ArrayDeque<>();
        boolean[] visited = new boolean[rooms.size()];

        stack.push(0);
        visited[0] = true;
        int visitedCount = 1;

        while (!stack.isEmpty()) {
            int room = stack.pop();

            for (int key : rooms.get(room)) {
                if (!visited[key]) {
                    visited[key] = true;
                    visitedCount++;
                    stack.push(key);
                }
            }
        }

        return visitedCount == rooms.size();
    }
}